### DISCUSSION TITLE
How to exclude whitespace from location information?

### DISCUSSION THREAD

**Discussion by radekstatebox** [peggyjs/peggy#228] [Q&A]
The usual way to ignore whitespace is by appending a whitespace rule to each terminal:

```pegjs
identifier = @$([a-z]+) _
_ = [ \t]*
```

However, now the location includes the whitespace:

```pegjs
Identifier = name:identifier { return { sourceSpan: location(), name } }
```

For the input `a b`, the location for the first identifier would be:

```js
{
    start: { offset: 0, ... },
    end: { offset: 2, ... }, // 2 rather than 1
}
```

Is there a nice way to exclude the trailing whitespace from the location information? I could move the `location()` call to the definition of `identifier` like so:

```pegjs
identifier = @(name:$([a-z]+) { return { sourceSpan: location(), name } }) _
```

But that doesn’t really scale to more complicated productions (especially non-terminals).

**Comment by hildjj** [comment]
Your suggestion might be cleaner once #337 lands.  Something like:

```peg.js
Trail_<R, n> = @(r:R { return { sourceSpan: location(), [n]: r } }) _
Identifier = Trail_<identifier, "name">
```

Except.  You'd really like access to the rule name of R so you didn't also have to pass in "name".   @quadristan I haven't fully internalized how templates work yet, do you have any insights?
